gapii/client/adb.go

```diff
@@ -98,24 +102,73 @@ func StartOrAttach(ctx context.Context, p *android.InstalledPackage, a *android.
 
 	log.I(ctx, "Forwarding")
 	if err := d.Forward(ctx, adb.TCPPort(port), adb.NamedAbstractSocket("gapii")); err != nil {
-		return nil, log.Err(ctx, err, "Setting up port forwarding")
+		return nil, log.Err(ctx, err, "Setting up port forwarding for gapii")
 	}
 
 	// FileDir may fail here. This happens if/when the app is non-debuggable.
 	// Don't set up vulkan tracing here, since the loader will not try and load the layer
 	// if we aren't debuggable regardless.
-	if err := d.Command("shell", "setprop", "debug.vulkan.layers", "VkGraphicsSpy").Run(ctx); err != nil {
-		// Clone context to ignore cancellation.
-		ctx := keys.Clone(context.Background(), ctx)
-		d.RemoveForward(ctx, port)
-		return nil, log.Err(ctx, err, "Setting up vulkan layer")
+	var vkDevLock *flock.DeviceFLock
+	if o.APIs&VulkanAPI != uint32(0) {
+		if err := d.SetSystemProperty(ctx, "debug.vulkan.layers", "VkGraphicsSpy"); err != nil {
+			// Clone context to ignore cancellation.
+			ctx := keys.Clone(context.Background(), ctx)
+			d.RemoveForward(ctx, port)
+			return nil, log.Err(ctx, err, "Setting up vulkan layer")
+		}
+		var err error
+		vkDevLock, err = flock.LockDevice(d.Instance().GetSerial())
+		if err != nil {
+			return nil, log.Err(ctx, err, "Reserve device for tracing Vulkan API")
+		}
+		// Make a thread to listen to the debug.vulkan.layers unset signal.
+		go func() {
+			port, err := adb.LocalFreeTCPPort()
+			if err != nil {
+				log.Err(ctx, err, "Finding free port for listening to debug.vulkan.layers unset signal.")
+				return
+			}
+			if err = d.Forward(ctx, adb.TCPPort(port), adb.NamedAbstractSocket("unset-debug-vulkan-layers")); err != nil {
+				log.Err(ctx, err, "Setting up port forwarding for listening to debug.vulkan.layers unset signal.")
+				return
+			}
+			defer d.RemoveForward(ctx, port)
+			// Try to connect to the unset signal port, if the port is open
+			// and messages can be received, unset the debug.vulkan.layers.
+			log.I(ctx, "Waitting for debug.vulkan.layers unset signal on localhost:%d", port)
+			task.Retry(ctx, 0, time.Second, func(ctx context.Context) (bool, error) {
+				conn, err := net.Dial("tcp", fmt.Sprintf("localhost:%d", port))
+				if err != nil {
+					return false, err
+				}
+				conn.SetReadDeadline(time.Now().Add(500 * time.Millisecond))
+				var n int
+				// No need to read the whole message
+				n, err = conn.Read(make([]byte, 1))
+				if err != nil || n == 0 {
+					return false, err
+				}
+				log.I(ctx, "VkGraphcsSpy loaded, unset debug.vulkan.layers now.")
+				if err = d.SetSystemProperty(ctx, "debug.vulkan.layers", ""); err != nil {
```

Comment:
I understand that this might be the best option we have right now, but I'd like us to consider moving GAPII and GAPIR over to gRPC this year. If we migrated to gRPC, unsetting the property could just be an RPC call and all of this stuff could be cleaned up.

Might be worth adding a comment.

Reply:
Done